File: src/main/java/com/refinedmods/refinedstorage/network/craftingmonitor/CraftingMonitorCancelMessage.java
package com.refinedmods.refinedstorage.network.craftingmonitor;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.container.CraftingMonitorContainerMenu;

import java.util.UUID;
import javax.annotation.Nullable;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.handling.PlayPayloadContext;

public class CraftingMonitorCancelMessage implements CustomPacketPayload {
    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "crafting_monitor_cancel");

    @Nullable
    private final UUID taskId;

    public CraftingMonitorCancelMessage(@Nullable UUID taskId) {
        this.taskId = taskId;
    }

    public static CraftingMonitorCancelMessage decode(FriendlyByteBuf buf) {
        return new CraftingMonitorCancelMessage(buf.readBoolean() ? buf.readUUID() : null);
    }

    public static void handle(CraftingMonitorCancelMessage message, PlayPayloadContext ctx) {
        ctx.player().ifPresent(player -> ctx.workHandler().submitAsync(() -> {
            if (player.containerMenu instanceof CraftingMonitorContainerMenu) {
                ((CraftingMonitorContainerMenu) player.containerMenu).getCraftingMonitor().onCancelled(
                    (ServerPlayer) player,
                    message.taskId
                );
            }
        }));
    }

    @Override
    public void write(FriendlyByteBuf buf) {
        buf.writeBoolean(taskId != null);

        if (taskId != null) {
            buf.writeUUID(taskId);
        }
    }

    @Override
    public ResourceLocation id() {
        return ID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/craftingmonitor/CraftingMonitorCancelMessage.kt
package com.refinedmods.refinedstorage.network.craftingmonitor

import com.refinedmods.refinedstorage.container.CraftingMonitorContainer
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import java.util.*
import java.util.function.Supplier

class CraftingMonitorCancelMessage(@field:Nullable @param:Nullable private val taskId: UUID?) {
    companion object {
        fun decode(buf: PacketByteBuf): CraftingMonitorCancelMessage {
            return CraftingMonitorCancelMessage(if (buf.readBoolean()) buf.readUniqueId() else null)
        }

        fun encode(message: CraftingMonitorCancelMessage, buf: PacketByteBuf) {
            buf.writeBoolean(message.taskId != null)
            if (message.taskId != null) {
                buf.writeUniqueId(message.taskId)
            }
        }

        fun handle(message: CraftingMonitorCancelMessage, ctx: Supplier<NetworkEvent.Context>) {
            val player: ServerPlayerEntity = ctx.get().getSender()
            if (player != null) {
                ctx.get().enqueueWork({
                    if (player.openContainer is CraftingMonitorContainer) {
                        (player.openContainer as CraftingMonitorContainer).craftingMonitor.onCancelled(player, message.taskId)
                    }
                })
            }
            ctx.get().setPacketHandled(true)
        }
    }
}